ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/classpathupdater/ClasspathUpdaterTest.java

```diff
@@ -50,6 +57,32 @@ public void testUpdateClasspath_CreateFromScratch() throws Exception {
 		runUpdateClasspathJob();
 		assertClasspathAttribute("JavaSE-17", IClasspathAttribute.MODULE, null, Boolean::valueOf);
 		assertClasspathAttribute("library.jar", IClasspathAttribute.TEST, null, Boolean::valueOf);
+		assertClasspathProperty("library.jar", "exported=true", (e) -> e.isExported());
+		assertClasspathAttribute("A.jar", IClasspathAttribute.TEST, null, Boolean::valueOf);
+		assertClasspathProperty("A.jar", "exported=true", (e) -> e.isExported());
+	}
+
+	@Test
+	public void testClasspathComputer_ChangeSourceAttachment() throws Exception {
+		assertTrue("Project was not created", project.exists());
+
+		Map<String, IPath> sourceMap = new HashMap<>();
+		sourceMap.put("A.jar", new Path("library.jar"));
+		sourceMap.put("library.jar", new Path("A.jar"));
+		IPluginModelBase model = PluginRegistry.findModel(project.getProject());
+
+		IClasspathEntry[] cp = ClasspathComputer.getClasspath(project, model, sourceMap, false, false);
+		jProject.setRawClasspath(cp, null);
+
+		assertClasspathAttribute("JavaSE-17", IClasspathAttribute.MODULE, true, Boolean::valueOf);
+		assertClasspathAttribute("library.jar", IClasspathAttribute.TEST, true, Boolean::valueOf);
+		assertClasspathProperty("library.jar", "exported=true", (e) -> e.isExported());
+		assertClasspathProperty("library.jar", "source",
+				(e) -> "A.jar".equals(nullOr(e.getSourceAttachmentPath(), IPath::lastSegment)));
```

Comment:
Somehow I failed to properly use java Optional here, hence the 'nullOr'. Any better idea from Java experts?